### PR TITLE
feat(client): read EMS rollup at detect time + sanity-check it

### LINKED ISSUE #95: EMS plant decoding follow-up: meter validity filter, IR(2040,55) verification, inverter cross-check
Follow-up to #86 / #93, opened after walking the wire capture from a real EMS install (see #86 for the captures and [dewet22/givenergy-hass#52](https://github.com/dewet22/givenergy-hass/issues/52) for the topology).

## Context

#93 stops the library issuing inverter-style reads against EMS plant controllers. The wire data also clarified something I'd had wrong: an EMS plant **doesn't** require sub-address inverter walking. Per-inverter telemetry is rolled up into the EMS's own `IR(2040..2094)` block (`inverter_N_power`, `inverter_N_soc`, `inverter_N_temp`, `inverter_N_serial_number`, plus per-meter and plant aggregates). Our `Ems` model already decodes all of it. So the remaining work after #93 is targeted, not architectural.

### 1. Filter empty meter slots in `detect()`

`Client.detect()`'s meter-probing loop appends any address in `0x01..0x08` that *responds* to `IR(60,30)` — but some EMS firmwares acknowledge unused slots with all-zero cache contents (see #86 capture: `0x07`, `0x08` returned zeros while `0x01`, `0x03` returned real voltage/current/power). Those zero-only slots become empty `Meter` objects in `plant.meters` for no benefit.

Suggested: after a probe response, gate the `caps.meter_addresses.append(...)` on `Meter.from_register_cache(cache).is_valid()` (which already checks `v_phase_1 ≠ 0`). Same model the library uses everywhere else; cheap and localised.

### 2. Wire-confirm the EMS `IR(2040,55)` read

#93 leaves the EMS-conditional `IR(2040,55)` append in place but documents it as unconfirmed — the existing capture ran with `capabilities=null`, so the read never fired. A 30-minute passive capture from the same hardware (requested over on hass#52) will let us decode the live values from `Ems.inverter_N_*` / `meter_N_*` against ground truth from the user's plant description (two Giv-AC 3.0 inverters, grid + load CTs, etc.) and either tighten the read range or extend it.

Not blocking — the read works in code today, just hasn't been exercised against hardware yet.

### 3. Cross-check `EMS.inverter_count` / `inverter_N_serial_number` against detect()

When the EMS reports `inverter_count = 2` with their serial numbers as 10-character strings, we could cross-reference those at detect time to:

- Validate the EMS register block parsed sensibly (count and serial-string charsets are easy sanity checks)
- Let downstream consumers (`givenergy-hass`) name entities by inverter serial rather than positional index

Optional — adds polish, not function.

## Out of scope here

- Integration-side sensor descriptors and config-entry UX — that's [dewet22/givenergy-hass#52](https://github.com/dewet22/givenergy-hass/issues/52)
- Multi-dongle / per-inverter modbus support — not needed for EMS plants, see "Context" above. If a different deployment topology surfaces this need later, it'd want its own issue.

### PR DESCRIPTION
**Pull request by dewet22**
Closes #95 — item 3 (items 1 and 2 already addressed in #96 and #95's wire-confirmation comment respectively).

## Summary

For EMS plant controllers, `Client.detect()` now issues an `IR(2040,55)` read so the per-managed-inverter and per-meter rollup is populated as part of discovery. The decoded rollup is sanity-checked:

- `inverter_count` must be in `[1, 4]`
- Each of the first `inverter_count` `inverter_N_serial_number` slots must match the GE 10-char format (`[A-Z]{2}\d{4}[A-Z]\d{3}`, which accepts both real and CLI-redacted forms)

Anomalies log warnings but never raise — discovery shouldn't fail on a soft data check. The intent is to surface parser regressions early and let downstream consumers (`givenergy-hass`) name entities by inverter serial rather than positional index without an extra refresh cycle.

## Wire-verified

Decoded against the EMS capture at `tests/fixtures/captures/ems_plant_a/ems_arm1036_30min.log` (committed in #103). The 30-minute window contains 82 `IR(2040,55)` responses; the existing `Ems` model decodes them cleanly and the cross-check passes against the contributing plant (`inverter_count = 2`, both serials match the documented `CE####G###` family redacted form).

Note: separate from this PR, the per-slot status bitfields (`IR(2043)` for meters, `IR(2045)` for inverters) in the existing `Ems` model don't match observed values on the wire — surfaced during this work and filed as [#108](https://github.com/dewet22/givenergy-modbus/issues/108) so it doesn't get lost. That's a data-integrity issue independent of the cross-check landing here.

## Test plan

- [x] `tests/client/test_detect.py::test_detect_ems_reads_rollup_at_detect_time` — asserts exactly one `IR(2040,55)` at device `0x11` per detect()
- [x] `tests/client/test_detect.py::test_detect_ems_warns_on_implausible_inverter_count` — primes `inverter_count = 7`, asserts warning
- [x] `tests/client/test_detect.py::test_detect_ems_warns_on_malformed_serial` — primes a junk serial in slot 1 and a valid one in slot 2, asserts warning only for slot 1
- [x] Existing `test_detect_ems_skips_lv_battery_probing` updated to prime the rollup too, so the warning path isn't accidentally triggered
- [x] All 30 detect-test cases pass
- [x] `uv run --group test tox` — full matrix green

## Summary by CodeRabbit

* **New Features**
  * EMS system discovery now validates inverter configuration data, including counts and serial numbers.
  * Anomalies detected during validation are logged as warnings and do not disrupt the discovery process.
  * Improved error detection and diagnostic logging for EMS-based installations.

* **Tests**
  * Expanded test suite with comprehensive coverage for EMS validation, including edge cases and malformed data scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/dewet22/givenergy-modbus/pull/109?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: givenergy_modbus/client/client.py [hunk 1]
```diff
@@ -425,6 +471,19 @@ async def detect(
                 ", ".join(f"0x{a:02x}" for a in caps.lv_battery_addresses),
             )
 
+        # Step 5 — EMS rollup cross-check. Read IR(2040,55) at detect time so the per-managed-
+        # inverter and per-meter rollup is populated, and sanity-check the parsed values to
+        # catch a malformed rollup (or a parser regression) early rather than letting consumers
+        # read garbage. Best-effort: log warnings on anomaly, never raise — discovery shouldn't
+        # fail on a soft data check. See #95.
+        if caps.is_ems:
+            await self.send_request_and_await_response(
```
Comment (dewet22) [inline givenergy_modbus/client/client.py]: This read still makes the EMS cross-check fatal if the rollup request times out or returns an error. The comment says the cross-check is best-effort and that discovery should not fail on it, but `send_request_and_await_response()` can raise `TimeoutError` before `_validate_ems_rollup()` gets a chance to log and continue. I would either wrap this read and log/continue on failure, or update the contract/tests to make the rollup read a required part of EMS discovery.

File: givenergy_modbus/client/client.py [hunk 2]
```diff
@@ -291,6 +298,45 @@ async def _detect_bcu_stacks(
...
+            )
+            return
+        serials = [ems.get(f"inverter_{i}_serial_number") for i in range(1, inverter_count + 1)]
+        for i, serial in enumerate(serials, start=1):
+            if not (isinstance(serial, str) and _GE_SERIAL_STR_PATTERN.fullmatch(serial)):
+                _logger.warning(
+                    "detect: EMS rollup inverter_%d_serial_number=%r doesn't match GE serial format",
+                    i,
+                    serial,
+                )
```
[the remark this answers is not in the document]
Comment (dewet22) [reply givenergy_modbus/client/client.py]: Applied — decoded serial is now stripped of trailing `\x00` and space padding before the GE-format check, so a padded-but-valid serial won't fire a false warning. No observable behaviour change against the current 5-register / 10-byte field shape, but defensive against a future wider layout. Pushed in commit 32950a4.

File: givenergy_modbus/client/client.py (hunk 1 above)
Comment (dewet22) [reply givenergy_modbus/client/client.py]: Good catch — pushed fix in commit db39fc7. The rollup read is now wrapped in try/except TimeoutError so the whole cross-check is genuinely best-effort. Extracted the block into `_ems_rollup_cross_check()` while there (keeps `detect()` under the McCabe threshold). New test `test_detect_ems_tolerates_rollup_read_timeout` covers the timeout path.

File: givenergy_modbus/client/client.py [hunk 3]
```diff
@@ -291,6 +298,45 @@ async def _detect_bcu_stacks(
...
+        _logger.info(
+            "detect: EMS rollup cross-check — inverter_count=%d, serials=[%s]",
+            inverter_count,
+            ", ".join(repr(s) for s in serials),
+        )
```
[the remark this answers is not in the document]
Comment (dewet22) [reply givenergy_modbus/client/client.py]: Applied — dropped the success log to DEBUG in commit 33a05ae. You're right that emitting full decoded serials at INFO is inconsistent with the redaction posture we've built up around `redact()` (#99); the per-slot WARNING still surfaces anomalies that need attention.
